### PR TITLE
indexing support (READY - no of lines failling)

### PR DESCRIPTION
**Pull request by kartik4949**

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -87,6 +87,11 @@ def __init__(self, data, gpu=None, requires_grad=True):
   def __repr__(self):
     return f"Tensor {self.data!r} with grad {(self.grad.data if self.grad else None)!r}"
 
+  def __getitem__(self, index):
+    if not isinstance(index, (int, list, tuple)):
+      raise ValueError("{type(index)} not a valid index.")
+    return self.data[index] if not self.gpu else self.cpu().data[index]
```
Comment (geohot) [inline tinygrad/tensor.py:93]: That can't convert it to cpu. If you want indexing, it has to work on the gpu too. It should also return Tensor type, not data.

Comment (kartik4949) [reply tinygrad/tensor.py:93]: Sorry i though i pushed that 
Pushing it

Comment (kartik4949) [reply tinygrad/tensor.py]: @geohot 
so for gpu
def gpu_index(index):
# Index GpuBuffer 
# return a new GPUBuffer with index data.
right?

**Comment by geohot** [comment]
Gonna need a whole bunch of tests for this too. This is a complex feature!

**Comment by kartik4949** [comment]
> Gonna need a whole bunch of tests for this too. This is a complex feature!

Yes, just added support for multidimensional too 
Will push with tests

**Comment by kartik4949** [comment]
@geohot please refactor if possible 
pr support 
- gpu single indexing i.e tensor[0]
- slicing indexing i.e tensor[2:10] NOTE: not supported steps yet.
- gpu multi dimensional indexing i.e tensor[2,4,5]

**Comment by kartik4949** [comment]
@geohot please refactor if possible 
pr is completed with support: 
- gpu single indexing i.e tensor[0]
- slicing indexing i.e tensor[2:10] NOTE: not supported steps yet.
- gpu multi dimensional indexing i.e tensor[2,4,5]

**Comment by kartik4949** [comment]
doesnt support slicing in multidimensional index, i.e tensor[1:2, 3:4, 5:7]

**Comment by kartik4949** [comment]
@geohot is this right? :)

**Comment by geohot** [comment]
I have no idea. Are the tests good and complete?

File: test/test_tensor.py [hunk 2]
```diff
@@ -66,6 +66,19 @@ def test_gradcheck(self):
     # coarse approx. since a "big" eps and the non-linearities of the model
     self.assertFalse(gradcheck(tiny_func, tiny_x, eps = 0.1))
 
+  def test_multi_indexing(self):
```
Comment (geohot) [inline test/test_tensor.py:69]: indentation. and why are the tests using randomness at all?

Comment (kartik4949) [reply test/test_tensor.py:69]: > I have no idea. Are the tests good and complete?

> indentation. and why are the tests using randomness at all?

for generating shape, and indexes ...

**Comment by kartik4949** [comment]
@geohot the algorithm is simple

[1,2,3]
[4,5,6] ->
 gpu buffer stores this as 

[1,2,3,4,5,6]

so i wrote an algorithm to access the correct element 👍🏽 
and tests are good i think 
for testing all features

**Comment by kartik4949** [comment]
> I have no idea. Are the tests good and complete?

you can do a sanity check too..pull up an interpreter and do some indexing.

File: test/test_tensor.py (hunk 2 above)
Comment (kartik4949) [reply test/test_tensor.py:69]: I can add hardcode the values, if u want.

**Comment by kartik4949** [comment]
Added more test cases,
and setup method, think how we can use it better.

**Comment by kartik4949** [comment]
Closing because couldnt find cleaner way to implement! may be some else can implement better!

**Pull request closed without merging**